src/Parser.elm

```diff
@@ -0,0 +1,374 @@
+                    , Jd.errorToString err
+                    ]
+
+        ( Ok folders, Ok audiences ) ->
+            let
+                clean =
+                    cleanUpData folders audiences
+            in
+            case checkData clean of
+                Nothing ->
+                    Ok clean
+
+                Just err ->
+                    Err err
+
+
+checkData : Data -> Maybe String
+checkData data =
+    maybeChain
+        [ allParentsAreFolders data
+        , allParentsExist data.all (Dict.toList data.parents)
+        ]
+
+
+allParentsExist : Dict.Dict Int String -> List ( Int, Int ) -> Maybe String
+allParentsExist all parents =
+    case parents of
+        [] ->
+            Nothing
+
+        ( child, parent ) :: arents ->
+            case Dict.get parent all of
+                Nothing ->
+                    Just <|
+                        String.concat
+                            [ "parent with ID "
+                            , String.fromInt parent
+                            , " and child "
+                            , String.fromInt child
+                            , " does not exist"
+                            ]
+
+                Just _ ->
+                    allParentsExist all arents
+
+
+allParentsAreFolders : Data -> Maybe String
+allParentsAreFolders { parents, audiences } =
+    allParentsAreFoldersHelp audiences (Dict.toList parents)
+
+
+allParentsAreFoldersHelp :
+    Dict.Dict Int A.AudienceType
+    -> List ( Int, Int )
+    -> Maybe String
+allParentsAreFoldersHelp audiences parents =
+    case parents of
+        [] ->
+            Nothing
+
+        ( child, parent ) :: arents ->
+            case Dict.get parent audiences of
+                Nothing ->
+                    allParentsAreFoldersHelp audiences arents
+
+                Just _ ->
+                    Just <|
+                        String.concat
+                            [ "parent with ID "
+                            , String.fromInt parent
+                            , " and child "
+                            , String.fromInt child
+                            , " is not a folder"
+                            ]
+
+
+maybeChain : List (Maybe a) -> Maybe a
+maybeChain maybes =
+    case maybes of
+        [] ->
+            Nothing
+
+        Nothing :: aybes ->
+            maybeChain aybes
+
+        (Just m) :: _ ->
+            Just m
+
+
+cleanUpData : List F.AudienceFolder -> List A.Audience -> Data
+cleanUpData folders audiences =
+    makeModel
+        { folders = folders
+        , audiences = audiences
+        , model = zeroDataSet
+        , unique = 0
+        , folderIds = Dict.empty
+        , audienceIds = Dict.empty
+        }
+
+
+type alias MakeModel =
+    { folders : List F.AudienceFolder
+    , audiences : List A.Audience
+    , model : Data
+    , unique : Int
+    , folderIds : Dict.Dict Int Int
+    , audienceIds : Dict.Dict Int Int
+    }
```

Comment:
Yes I agree. I have now changed it so that the fields are accumulated separately, and it's also much simpler now it doesn't calculate new IDs, following your suggestion above.